### PR TITLE
[PM-7926] Handle complex user logout events

### PR DESCRIPTION
**Pull request by MGibson1**
## Type of change

```
- [x] Bug fix
- [ ] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

This eneded up being a bit of a hodge-podge of fixing issues around logging our and nextup user determination.

1. Browser now switches accounts first (including no account)
   - the switch account method redirects when necessary
2. Split app logout responsibilities to showing a toast and navigating home. Home is handled by switch and toast by logout. This is why we can always send `"doneLoggingOut"`
3. Chrome, Vivaldi, and Opera all reseed storage on logout, which forces the disk database log to clear ([background](https://github.com/bitwarden/clients/issues/485)). These reseed events were picked up as `remove` and `save` events that caused numerous race conditions on logout with components and services still using `firstValueFrom`. Ignoring events that don't match the signature of our storage services -- and therefore are special cases not due to the state provider framework -- resolves all the races I was seeing around logout.
4. Activity is set on account switch, which results in correctly determining the next up user even when the activity heartbeat has not yet thumped
5. Cipher Service has been updated to support user data clearing by userId. Individual updates still assume active user, but the interface used by logout / lock now supports passing in a userId so we can perform these actions appropriately on non-active users.

## Code changes

- **account-switcher.component**: Navigation now occurs in background + app component through messaging, it is inappropriate to trigger navigation manually here because we lack context on next up users.
- **main.background**:
  - await changes
  - popup-only (and therefore null in popup) service handling in lock/logout methods
  - early switch and improved switch navigation routing
  - support displaying expired toast on switch, not just logout to unauth
- **abstract-chrome-storage-api**: ignore updates that don't look like storage service-mediated storage changes.
- **app.component**: handle routing for no next up user
- **account service**: trigger activity update on switch
- **cipher service**: handle non active data clear

## Before you submit

- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team
- Ensure that all UI additions follow [WCAG AA requirements](https://contributing.bitwarden.com/contributing/accessibility/)

### CODE REVIEW

File: apps/browser/src/platform/services/abstractions/abstract-chrome-storage-api.service.ts [hunk 1]
```diff
@@ -34,6 +34,11 @@ export default abstract class AbstractChromeStorageService
 
   constructor(protected chromeStorageApi: chrome.storage.StorageArea) {
     this.updates$ = fromChromeEvent(this.chromeStorageApi.onChanged).pipe(
+      filter(([changes]) => {
+        // Our storage services support changing only one key at a time. If more are changed, it's due to
+        // reseeding storage and we should ignore the changes.
```
Comment (cagonzalezcs) [inline apps/browser/src/platform/services/abstractions/abstract-chrome-storage-api.service.ts:39]: Hm, are these changes just swallowed and ignored entirely?

Is there any change that a set of legitimate changes could occur simultaneously? Would we miss updating local storage correctly in that case?

File: apps/browser/src/background/main.background.ts [hunk 2]
```diff
@@ -360,9 +360,7 @@ export default class MainBackground {
     // Services
     const lockedCallback = async (userId?: string) => {
       if (this.notificationsService != null) {
-        // FIXME: Verify that this floating promise is intentional. If it is, add an explanatory comment and ensure there is proper error handling.
-        // eslint-disable-next-line @typescript-eslint/no-floating-promises
-        this.notificationsService.updateConnection(false);
+        void this.notificationsService.updateConnection(false);
```
Comment (cagonzalezcs) [inline apps/browser/src/background/main.background.ts:363]: Your notes state that we should no longer be triggering lock/logout/account switch behavior within the popupOnly context... I'm not seeing anything here that might guard against that, is that the case or am I missing something?

File: apps/browser/src/platform/services/abstractions/abstract-chrome-storage-api.service.ts (hunk 1 above)
Comment (MGibson1) [reply apps/browser/src/platform/services/abstractions/abstract-chrome-storage-api.service.ts:39]: Any changes to storage that are done through our storage framework and not directly accessing the storage apis will not be swallowed by this.

I toyed around with creating a property that could be synced and indicate that the service should ignore updates while it is true, but that causes concurrency issues.

The only other idea I have is to write/remove a special value and any changeset involving that value would be ignored by this update stream. I kind of like that better now that I say it.

File: apps/browser/src/background/main.background.ts (hunk 2 above)
Comment (MGibson1) [reply apps/browser/src/background/main.background.ts:363]: Do you mean in this file or _right_ here? This specific section is just getting rid of a `FIXME`.

I also didn't mean to imply that there is no direct call to lock or logout from a popup, just that if there is, unintialized services are through optional chaining operators.

Comment (cagonzalezcs) [reply apps/browser/src/background/main.background.ts:363]: I see, ok that makes sense.

File: apps/browser/src/platform/services/abstractions/abstract-chrome-storage-api.service.ts (hunk 1 above)
Comment (cagonzalezcs) [reply apps/browser/src/platform/services/abstractions/abstract-chrome-storage-api.service.ts:39]: Maybe worth exploring in a future refactor.

**Review by cagonzalezcs — DISMISSED** [review]

**Review by Jingo88 — DISMISSED** [review]

File: apps/browser/src/platform/services/abstractions/abstract-chrome-storage-api.service.ts (hunk 1 above)
Comment (MGibson1) [reply apps/browser/src/platform/services/abstractions/abstract-chrome-storage-api.service.ts:39]: looks like that does work. I don't think I like it more, though. It means that we catch non-storage service compliant changes, but it also means that if there are any more weird updates like reseed, we need to identify them manually. So the question is: Do we think these kind of out-of-band updates should be generally honored or ignored? My instinct is ignored.

**Review by jlf0dev — APPROVED** [review]
We need tests _really_ badly for this process. I would be scared to touch any of this code.

File: apps/browser/src/background/main.background.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (MGibson1) [reply apps/browser/src/background/main.background.ts]: reverted

**Review by cagonzalezcs — APPROVED** [review]

File: libs/angular/src/services/jslib-services.module.ts [hunk 3]
```diff
@@ -629,6 +629,7 @@ const safeProviders: SafeProvider[] = [
       LOGOUT_CALLBACK,
       BillingAccountProfileStateService,
       TokenServiceAbstraction,
+      AuthService,
```
Comment (justindbaur) [inline libs/angular/src/services/jslib-services.module.ts:632]: This needs to have the `Abstraction` suffix.

```suggestion
      AuthServiceAbstraction,
```

**Review by justindbaur — CHANGES_REQUESTED** [review]
One tiny thing.

**Review by Jingo88 — DISMISSED** [review]

**Review by jlf0dev — DISMISSED** [review]

File: libs/angular/src/services/jslib-services.module.ts (hunk 3 above)
Comment (MGibson1) [reply libs/angular/src/services/jslib-services.module.ts]: good catch, thank you

**Review by justindbaur — APPROVED** [review]

**Review by jprusik — APPROVED** [review]
Smoke tested and LGTM

**Review by Jingo88 — APPROVED** [review]

**Review by jlf0dev — APPROVED** [review]

**Pull request merged**